{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68a4b4df",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import tracker\n",
    "latest_date = tracker.processing_dates[-1]\n",
    "os.chdir('/mnt/BioAdHoc/Groups/vd-ay/kfetter/hichip-db-loop-calling/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d83566ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2022.06.27.09.17'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "latest_date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8f649648",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting input and output with jupyter notebook in context \n",
    "if 'ipykernel_launcher.py' in sys.argv[0]:\n",
    "    input_fn = 'results/samplesheets/fastq/{}.fastq.google-samplesheet.tsv'.format(latest_date)\n",
    "    output_prefix = 'results/samplesheets/fastq/{}.fastq.samplesheet'.format(latest_date)\n",
    "# if not using jupyter notebook\n",
    "else:\n",
    "    input_fn = sys.argv[1]\n",
    "    output_prefix = sys.argv[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a054afb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading the samplesheet\n",
    "df = pd.read_table(input_fn, skiprows=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01461a69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name (as used in the server)</th>\n",
       "      <th>Study</th>\n",
       "      <th>GSE ID</th>\n",
       "      <th>GSM ID</th>\n",
       "      <th>SRR ID</th>\n",
       "      <th>GEO Title</th>\n",
       "      <th>GEO Source</th>\n",
       "      <th>GEO Description</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Number of Reads</th>\n",
       "      <th>...</th>\n",
       "      <th>HiCPro Status</th>\n",
       "      <th>HiChIP Peaks Status</th>\n",
       "      <th>hichipper Status</th>\n",
       "      <th>OTHER Status</th>\n",
       "      <th>FitHiChIP Status</th>\n",
       "      <th>Operator</th>\n",
       "      <th>Priority</th>\n",
       "      <th>Cluster Path/Branch</th>\n",
       "      <th>Date Added</th>\n",
       "      <th>Comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GM</td>\n",
       "      <td>Mumbach et al., 2017</td>\n",
       "      <td>GSE101498</td>\n",
       "      <td>GSM2705041</td>\n",
       "      <td>SRR5831489</td>\n",
       "      <td>GM HiChIP H3K27ac biological replicate 1</td>\n",
       "      <td>GM12878 cell line</td>\n",
       "      <td>Protein-enriched long-range contact; GM_HiChIP...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>664,558,514</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pre-5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM</td>\n",
       "      <td>Mumbach et al., 2017</td>\n",
       "      <td>GSE101498</td>\n",
       "      <td>GSM2705042</td>\n",
       "      <td>SRR5831490</td>\n",
       "      <td>GM HiChIP H3K27ac biological replicate 2</td>\n",
       "      <td>GM12878 cell line</td>\n",
       "      <td>Protein-enriched long-range contact; GM_HiChIP...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>598,957,472</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pre-5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K562</td>\n",
       "      <td>Mumbach et al., 2017</td>\n",
       "      <td>GSE101498</td>\n",
       "      <td>GSM2705043</td>\n",
       "      <td>SRR5831491</td>\n",
       "      <td>K562 HiChIP H3K27ac biological replicate 1</td>\n",
       "      <td>K562 cell line</td>\n",
       "      <td>Protein-enriched long-range contact; K562_HiCh...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>426,330,778</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pre-5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K562</td>\n",
       "      <td>Mumbach et al., 2017</td>\n",
       "      <td>GSE101498</td>\n",
       "      <td>GSM2705044</td>\n",
       "      <td>SRR5831492</td>\n",
       "      <td>K562 HiChIP H3K27ac biological replicate 2</td>\n",
       "      <td>K562 cell line</td>\n",
       "      <td>Protein-enriched long-range contact; K562_HiCh...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>311,815,900</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pre-5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K562</td>\n",
       "      <td>Mumbach et al., 2017</td>\n",
       "      <td>GSE101498</td>\n",
       "      <td>GSM2705045</td>\n",
       "      <td>SRR5831493</td>\n",
       "      <td>K562 HiChIP H3K27ac biological replicate 3</td>\n",
       "      <td>K562 cell line</td>\n",
       "      <td>Protein-enriched long-range contact; K562_HiCh...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>356,325,902</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pre-5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>K562</td>\n",
       "      <td>Weintraub et al., 2017</td>\n",
       "      <td>GSE99519</td>\n",
       "      <td>GSM2774002</td>\n",
       "      <td>SRR6010263</td>\n",
       "      <td>HiChIP_K562_YY1</td>\n",
       "      <td>HiChIP_K562_YY1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>158111028</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>iPSC-WT</td>\n",
       "      <td>Kraft et al., 2022</td>\n",
       "      <td>GSE150906</td>\n",
       "      <td>GSM4560486</td>\n",
       "      <td>SRR11816734</td>\n",
       "      <td>iPSC_WT_A_H3K27me3_HiChIP</td>\n",
       "      <td>wild-type iPSC</td>\n",
       "      <td>H3K27me3 HiChIP library; iPSC_wt6_K1-TAAGGCGA-...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>195443840</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>iPSC-WT</td>\n",
       "      <td>Kraft et al., 2022</td>\n",
       "      <td>GSE150906</td>\n",
       "      <td>GSM4560487</td>\n",
       "      <td>SRR11816735</td>\n",
       "      <td>iPSC_WT_B_H3K27me3_HiChIP</td>\n",
       "      <td>wild-type iPSC</td>\n",
       "      <td>H3K27me3 HiChIP library; iPSC_wt7_K3-AGGCAGAA-...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>28509234</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>iPSC-MUT</td>\n",
       "      <td>Kraft et al., 2022</td>\n",
       "      <td>GSE150906</td>\n",
       "      <td>GSM4560488</td>\n",
       "      <td>SRR11816736</td>\n",
       "      <td>iPSC_MUT_A_H3K27me3_HiChIP</td>\n",
       "      <td>RNA-binding deficient EZH2 mutant iPSC</td>\n",
       "      <td>H3K27me3 HiChIP library; iPSC_mut18_K5-GGACTCC...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>309053672</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>iPSC-MUT</td>\n",
       "      <td>Kraft et al., 2022</td>\n",
       "      <td>GSE150906</td>\n",
       "      <td>GSM4560489</td>\n",
       "      <td>SRR11816737</td>\n",
       "      <td>iPSC_MUT_B_H3K27me3_HiChIP</td>\n",
       "      <td>RNA-binding deficient EZH2 mutant iPSC</td>\n",
       "      <td>H3K27me3 HiChIP library; iPSC_mut20_K7-CTCTCTA...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>112548012</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Joaquin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6/16/2022</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>485 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample Name (as used in the server)                   Study     GSE ID  \\\n",
       "0                                    GM    Mumbach et al., 2017  GSE101498   \n",
       "1                                    GM    Mumbach et al., 2017  GSE101498   \n",
       "2                                  K562    Mumbach et al., 2017  GSE101498   \n",
       "3                                  K562    Mumbach et al., 2017  GSE101498   \n",
       "4                                  K562    Mumbach et al., 2017  GSE101498   \n",
       "..                                  ...                     ...        ...   \n",
       "480                                K562  Weintraub et al., 2017   GSE99519   \n",
       "481                             iPSC-WT      Kraft et al., 2022  GSE150906   \n",
       "482                             iPSC-WT      Kraft et al., 2022  GSE150906   \n",
       "483                            iPSC-MUT      Kraft et al., 2022  GSE150906   \n",
       "484                            iPSC-MUT      Kraft et al., 2022  GSE150906   \n",
       "\n",
       "         GSM ID       SRR ID                                   GEO Title  \\\n",
       "0    GSM2705041   SRR5831489    GM HiChIP H3K27ac biological replicate 1   \n",
       "1    GSM2705042   SRR5831490    GM HiChIP H3K27ac biological replicate 2   \n",
       "2    GSM2705043   SRR5831491  K562 HiChIP H3K27ac biological replicate 1   \n",
       "3    GSM2705044   SRR5831492  K562 HiChIP H3K27ac biological replicate 2   \n",
       "4    GSM2705045   SRR5831493  K562 HiChIP H3K27ac biological replicate 3   \n",
       "..          ...          ...                                         ...   \n",
       "480  GSM2774002   SRR6010263                             HiChIP_K562_YY1   \n",
       "481  GSM4560486  SRR11816734                   iPSC_WT_A_H3K27me3_HiChIP   \n",
       "482  GSM4560487  SRR11816735                   iPSC_WT_B_H3K27me3_HiChIP   \n",
       "483  GSM4560488  SRR11816736                  iPSC_MUT_A_H3K27me3_HiChIP   \n",
       "484  GSM4560489  SRR11816737                  iPSC_MUT_B_H3K27me3_HiChIP   \n",
       "\n",
       "                                 GEO Source  \\\n",
       "0                         GM12878 cell line   \n",
       "1                         GM12878 cell line   \n",
       "2                            K562 cell line   \n",
       "3                            K562 cell line   \n",
       "4                            K562 cell line   \n",
       "..                                      ...   \n",
       "480                         HiChIP_K562_YY1   \n",
       "481                          wild-type iPSC   \n",
       "482                          wild-type iPSC   \n",
       "483  RNA-binding deficient EZH2 mutant iPSC   \n",
       "484  RNA-binding deficient EZH2 mutant iPSC   \n",
       "\n",
       "                                       GEO Description      Organism  \\\n",
       "0    Protein-enriched long-range contact; GM_HiChIP...  Homo sapiens   \n",
       "1    Protein-enriched long-range contact; GM_HiChIP...  Homo sapiens   \n",
       "2    Protein-enriched long-range contact; K562_HiCh...  Homo sapiens   \n",
       "3    Protein-enriched long-range contact; K562_HiCh...  Homo sapiens   \n",
       "4    Protein-enriched long-range contact; K562_HiCh...  Homo sapiens   \n",
       "..                                                 ...           ...   \n",
       "480                                                NaN  Homo sapiens   \n",
       "481  H3K27me3 HiChIP library; iPSC_wt6_K1-TAAGGCGA-...  Homo sapiens   \n",
       "482  H3K27me3 HiChIP library; iPSC_wt7_K3-AGGCAGAA-...  Homo sapiens   \n",
       "483  H3K27me3 HiChIP library; iPSC_mut18_K5-GGACTCC...  Homo sapiens   \n",
       "484  H3K27me3 HiChIP library; iPSC_mut20_K7-CTCTCTA...  Homo sapiens   \n",
       "\n",
       "    Number of Reads  ...  HiCPro Status  HiChIP Peaks Status hichipper Status  \\\n",
       "0       664,558,514  ...              1                    0                0   \n",
       "1       598,957,472  ...              1                    0                0   \n",
       "2       426,330,778  ...              1                    0                0   \n",
       "3       311,815,900  ...              1                    0                0   \n",
       "4       356,325,902  ...              1                    0                0   \n",
       "..              ...  ...            ...                  ...              ...   \n",
       "480       158111028  ...              1                    0                0   \n",
       "481       195443840  ...              1                    0                0   \n",
       "482        28509234  ...              1                    0                0   \n",
       "483       309053672  ...              1                    0                0   \n",
       "484       112548012  ...              1                    0                0   \n",
       "\n",
       "     OTHER Status  FitHiChIP Status  Operator Priority  Cluster Path/Branch  \\\n",
       "0               0                 0   Joaquin      NaN                  NaN   \n",
       "1               0                 0   Joaquin      NaN                  NaN   \n",
       "2               0                 0   Joaquin      NaN                  NaN   \n",
       "3               0                 0   Joaquin      NaN                  NaN   \n",
       "4               0                 0   Joaquin      NaN                  NaN   \n",
       "..            ...               ...       ...      ...                  ...   \n",
       "480             0                 0   Joaquin      NaN                  NaN   \n",
       "481             0                 0   Joaquin      NaN                  NaN   \n",
       "482             0                 0   Joaquin      NaN                  NaN   \n",
       "483             0                 0   Joaquin      NaN                  NaN   \n",
       "484             0                 0   Joaquin      NaN                  NaN   \n",
       "\n",
       "        Date Added  Comments  \n",
       "0    pre-5/16/2022       NaN  \n",
       "1    pre-5/16/2022       NaN  \n",
       "2    pre-5/16/2022       NaN  \n",
       "3    pre-5/16/2022       NaN  \n",
       "4    pre-5/16/2022       NaN  \n",
       "..             ...       ...  \n",
       "480      5/16/2022       NaN  \n",
       "481      6/16/2022       NaN  \n",
       "482      6/16/2022       NaN  \n",
       "483      6/16/2022       NaN  \n",
       "484      6/16/2022       NaN  \n",
       "\n",
       "[485 rows x 26 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d8351f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract those samples which are ready for processing\n",
    "ready_df = df.loc[(df['Start Processing'] == 1) & (df['Download Status'] == 0) & (df['Operator'] == \"Kyra\")]\n",
    "\n",
    "# extract only the columns needed\n",
    "major_cols = ['Sample Name (as used in the server)',\n",
    " 'GSE ID',\n",
    " 'GSM ID',\n",
    " 'SRR ID',\n",
    " 'Organism',\n",
    " 'Biological Replicate Serial No',\n",
    " 'Technical Replicate Serial No',\n",
    " 'ChIP-seq Pull Down',\n",
    " 'Restriction Enzyme']\n",
    "ready_df = ready_df[major_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "97e6eae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# capitalizes organism\n",
    "def parse_organism(string):\n",
    "    new_words = []\n",
    "    for word in string.split():\n",
    "        new_words.append(word.capitalize())\n",
    "    new_string = '_'.join(new_words)\n",
    "    return(new_string)\n",
    "\n",
    "ready_df.loc[:, 'Organism'] = ready_df.loc[:, 'Organism'].apply(parse_organism)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f751ffc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the sample names\n",
    "sample_names = []\n",
    "for i, sr in ready_df.iterrows():\n",
    "    sample_name = '{sample_name}.{gse_id}.{organism}.{antibody_target}.b{biological_rep}'\n",
    "    sample_name = sample_name.format(sample_name=sr[0],\n",
    "                                     gse_id=sr[1],\n",
    "                                     organism=sr[4],\n",
    "                                     antibody_target=sr[7], \n",
    "                                     biological_rep=sr[5])\n",
    "    sample_names.append(sample_name)\n",
    "ready_df.loc[:, 'sample_name'] = sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4bc55019",
   "metadata": {},
   "outputs": [],
   "source": [
    "# renaming the columns for easy computational use \n",
    "ready_df.columns = ['sample_name', 'gse_id', 'gsm_id', 'srr_id',\n",
    "                    'organism', 'bio_rep', 'tech_rep', 'antibody_target',\n",
    "                    'restriction_enzyme', 'std_sample_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "04f07c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reorder the columns\n",
    "reorder = ['std_sample_name',\n",
    "             'gse_id',\n",
    "             'gsm_id',\n",
    "             'srr_id',\n",
    "             'organism',\n",
    "             'bio_rep',\n",
    "             'tech_rep',\n",
    "             'antibody_target',\n",
    "             'restriction_enzyme',\n",
    "             'sample_name']\n",
    "ready_df = ready_df[reorder]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "95a3a12f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>std_sample_name</th>\n",
       "      <th>gse_id</th>\n",
       "      <th>gsm_id</th>\n",
       "      <th>srr_id</th>\n",
       "      <th>organism</th>\n",
       "      <th>bio_rep</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>antibody_target</th>\n",
       "      <th>restriction_enzyme</th>\n",
       "      <th>sample_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>A673_SA1m1.GSE133227.Homo_Sapiens.CTCF.b1</td>\n",
       "      <td>GSE133227</td>\n",
       "      <td>GSM3902791</td>\n",
       "      <td>SRR9590180</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>MboI</td>\n",
       "      <td>A673_SA1m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>A673_SA1m1.GSE133227.Homo_Sapiens.CTCF.b2</td>\n",
       "      <td>GSE133227</td>\n",
       "      <td>GSM3902792</td>\n",
       "      <td>SRR9590181</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>MboI</td>\n",
       "      <td>A673_SA1m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b1</td>\n",
       "      <td>GSE133227</td>\n",
       "      <td>GSM3902793</td>\n",
       "      <td>SRR9590182</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>MboI</td>\n",
       "      <td>A673_SA2m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b2</td>\n",
       "      <td>GSE133227</td>\n",
       "      <td>GSM3902794</td>\n",
       "      <td>SRR9590183</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>MboI</td>\n",
       "      <td>A673_SA2m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b3</td>\n",
       "      <td>GSE133227</td>\n",
       "      <td>GSM3902795</td>\n",
       "      <td>SRR9590184</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>MboI</td>\n",
       "      <td>A673_SA2m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>DND41.GSE165207.Homo_Sapiens.H3K27ac.b2</td>\n",
       "      <td>GSE165207</td>\n",
       "      <td>GSM5028230</td>\n",
       "      <td>SRR13492047</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>Arima</td>\n",
       "      <td>DND41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>DND41.GSE165207.Homo_Sapiens.H3K27ac.b2</td>\n",
       "      <td>GSE165207</td>\n",
       "      <td>GSM5028230</td>\n",
       "      <td>SRR13492048</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>Arima</td>\n",
       "      <td>DND41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>Jurkat.GSE165207.Homo_Sapiens.H3K27ac.b1</td>\n",
       "      <td>GSE165207</td>\n",
       "      <td>GSM5028231</td>\n",
       "      <td>SRR13492049</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>Arima</td>\n",
       "      <td>Jurkat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>Jurkat.GSE165207.Homo_Sapiens.H3K27ac.b1</td>\n",
       "      <td>GSE165207</td>\n",
       "      <td>GSM5028231</td>\n",
       "      <td>SRR13492050</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>Arima</td>\n",
       "      <td>Jurkat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>CD34+-Cord-Blood.GSE165207.Homo_Sapiens.H3K27a...</td>\n",
       "      <td>GSE165207</td>\n",
       "      <td>GSM5028232</td>\n",
       "      <td>SRR13492051</td>\n",
       "      <td>Homo_Sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>Arima</td>\n",
       "      <td>CD34+-Cord-Blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>103 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       std_sample_name     gse_id      gsm_id  \\\n",
       "37           A673_SA1m1.GSE133227.Homo_Sapiens.CTCF.b1  GSE133227  GSM3902791   \n",
       "38           A673_SA1m1.GSE133227.Homo_Sapiens.CTCF.b2  GSE133227  GSM3902792   \n",
       "39           A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b1  GSE133227  GSM3902793   \n",
       "40           A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b2  GSE133227  GSM3902794   \n",
       "41           A673_SA2m1.GSE133227.Homo_Sapiens.CTCF.b3  GSE133227  GSM3902795   \n",
       "..                                                 ...        ...         ...   \n",
       "411            DND41.GSE165207.Homo_Sapiens.H3K27ac.b2  GSE165207  GSM5028230   \n",
       "412            DND41.GSE165207.Homo_Sapiens.H3K27ac.b2  GSE165207  GSM5028230   \n",
       "413           Jurkat.GSE165207.Homo_Sapiens.H3K27ac.b1  GSE165207  GSM5028231   \n",
       "414           Jurkat.GSE165207.Homo_Sapiens.H3K27ac.b1  GSE165207  GSM5028231   \n",
       "415  CD34+-Cord-Blood.GSE165207.Homo_Sapiens.H3K27a...  GSE165207  GSM5028232   \n",
       "\n",
       "          srr_id      organism  bio_rep  tech_rep antibody_target  \\\n",
       "37    SRR9590180  Homo_Sapiens        1         1            CTCF   \n",
       "38    SRR9590181  Homo_Sapiens        2         1            CTCF   \n",
       "39    SRR9590182  Homo_Sapiens        1         1            CTCF   \n",
       "40    SRR9590183  Homo_Sapiens        2         1            CTCF   \n",
       "41    SRR9590184  Homo_Sapiens        3         1            CTCF   \n",
       "..           ...           ...      ...       ...             ...   \n",
       "411  SRR13492047  Homo_Sapiens        2         1         H3K27ac   \n",
       "412  SRR13492048  Homo_Sapiens        2         2         H3K27ac   \n",
       "413  SRR13492049  Homo_Sapiens        1         1         H3K27ac   \n",
       "414  SRR13492050  Homo_Sapiens        1         2         H3K27ac   \n",
       "415  SRR13492051  Homo_Sapiens        1         1         H3K27ac   \n",
       "\n",
       "    restriction_enzyme       sample_name  \n",
       "37                MboI        A673_SA1m1  \n",
       "38                MboI        A673_SA1m1  \n",
       "39                MboI        A673_SA2m1  \n",
       "40                MboI        A673_SA2m1  \n",
       "41                MboI        A673_SA2m1  \n",
       "..                 ...               ...  \n",
       "411              Arima             DND41  \n",
       "412              Arima             DND41  \n",
       "413              Arima            Jurkat  \n",
       "414              Arima            Jurkat  \n",
       "415              Arima  CD34+-Cord-Blood  \n",
       "\n",
       "[103 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ready_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dc2ac328",
   "metadata": {},
   "outputs": [],
   "source": [
    "header_output = '{}.with_header.tsv'.format(output_prefix)\n",
    "ready_df.to_csv(header_output, header=True, index=False, sep='\\t')\n",
    "\n",
    "without_header_output = '{}.without_header.tsv'.format(output_prefix)\n",
    "ready_df.to_csv(without_header_output, header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "df015ee3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(485, 26)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8a5406fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(103, 10)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ready_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4711bd3a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hichip-db",
   "language": "python",
   "name": "hichip-db"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
