{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many reads\n",
    "# whether there was a control or not and whether it is used in peak calling\n",
    "# how many peaks at what setting and threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1493909331.py, line 40)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [14]\u001b[0;36m\u001b[0m\n\u001b[0;31m    if line['Merged replicates (if there are replicates)']==\u001b[0m\n\u001b[0m                                                            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "def generate_summary_df(path):\n",
    "    samples = []\n",
    "    with open(path,'r') as file:\n",
    "\n",
    "        #reader = csv.reader(file, delimiter='\\t')\n",
    "        #next(reader, None)\n",
    "        reader = csv.DictReader(file, delimiter='\\t')\n",
    "        for line in reader:\n",
    "            sample_info = []\n",
    "            \n",
    "            # if there is no sample name, skip this line\n",
    "            if not line['Sample Name']:\n",
    "                continue\n",
    "\n",
    "            ##### CONSTRUCT THE SAMPLE NAME ######\n",
    "\n",
    "            name = line['Sample Name']\n",
    "            gse = line['GSE ID of Corresponding HiChIP Data']\n",
    "\n",
    "            organism = line['Organism']\n",
    "            organism = string.capwords(organism).replace(\" \", \"_\")\n",
    "\n",
    "            target_of_antibody = line['ChIP-seq Pull Down']\n",
    "            if (target_of_antibody==\"N/A\"):\n",
    "                continue\n",
    "            \n",
    "            replicate = line['Replicate Serial No']\n",
    "\n",
    "            sample_name = name + \".\" + gse + \".\" + organism + \".\" + target_of_antibody + \".b\" + replicate\n",
    "\n",
    "            ###########\n",
    "\n",
    "            # determine whether a control is present\n",
    "            control_info = \"no control for this sample\"\n",
    "            if line['Control GSM(s)']!=\"N/A\":\n",
    "                control_info = \"control present and used in peak calling\"\n",
    "\n",
    "            \n",
    "            # determine peak sizes and settings\n",
    "            narrowpeak1 = \"peak calling not completed\"\n",
    "            narrowpeak2 = \"peak calling not completed\"\n",
    "\n",
    "            idrpeak1 = \"peak calling not completed\"\n",
    "            idrpeak2 = \"peak calling not completed\"\n",
    "            if line['Merged replicates (if there are replicates)']==\"0\":\n",
    "                pass\n",
    "\n",
    "            sample_info = [sample_name, control_info]\n",
    "            samples.append(sample_info)\n",
    "    \n",
    "    df = pd.DataFrame(samples)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Control Used?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EryP.GSE112717.Mus_Musculus.Gata1.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EryD.GSE112717.Mus_Musculus.Gata1.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MLL-AF9.GSE110898.Mus_Musculus.H3K27ac.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MLL-AF9-LSD1i.GSE110898.Mus_Musculus.H3K27ac.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BL6_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BL6_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BL6_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BL6_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>F1_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>F1_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NOD_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NOD_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NOD_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NOD_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Erythroid_0hr.GSE142004.Mus_Musculus.YY1.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Erythroid_0hr.GSE142004.Mus_Musculus.YY1.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Erythroid_1hr.GSE142004.Mus_Musculus.YY1.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Erythroid_1hr.GSE142004.Mus_Musculus.YY1.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>mESC_RYBP_KO.GSE147919.Mus_Musculus.RYBP.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BMDM_24hIL4.GSE159629.Mus_Musculus.H3K4me3.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BMDM_24hIL4.GSE159629.Mus_Musculus.H3K4me3.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BMDM_Basal.GSE159629.Mus_Musculus.H3K4me3.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BMDM_Basal.GSE159629.Mus_Musculus.H3K4me3.b2</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3T3_Norm.GSE178344.Mus_Musculus.SMC1A.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>3T3_TCF1.GSE178344.Mus_Musculus.SMC1A.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Sample Name  \\\n",
       "2              EryP.GSE112717.Mus_Musculus.Gata1.b1   \n",
       "3              EryD.GSE112717.Mus_Musculus.Gata1.b1   \n",
       "4    AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b1   \n",
       "5    AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b2   \n",
       "6    AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b1   \n",
       "7    AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b2   \n",
       "8         MLL-AF9.GSE110898.Mus_Musculus.H3K27ac.b1   \n",
       "9   MLL-AF9-LSD1i.GSE110898.Mus_Musculus.H3K27ac.b1   \n",
       "10      BL6_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1   \n",
       "11      BL6_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2   \n",
       "12    BL6_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b1   \n",
       "13    BL6_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b2   \n",
       "14       F1_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1   \n",
       "15       F1_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2   \n",
       "16      NOD_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b1   \n",
       "17      NOD_CD4_CD8.GSE141847.Mus_Musculus.SMC1A.b2   \n",
       "18    NOD_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b1   \n",
       "19    NOD_CD4_CD8.GSE141847.Mus_Musculus.H3K27ac.b2   \n",
       "20      Erythroid_0hr.GSE142004.Mus_Musculus.YY1.b1   \n",
       "21      Erythroid_0hr.GSE142004.Mus_Musculus.YY1.b2   \n",
       "22      Erythroid_1hr.GSE142004.Mus_Musculus.YY1.b1   \n",
       "23      Erythroid_1hr.GSE142004.Mus_Musculus.YY1.b2   \n",
       "24      mESC_RYBP_KO.GSE147919.Mus_Musculus.RYBP.b1   \n",
       "25    BMDM_24hIL4.GSE159629.Mus_Musculus.H3K4me3.b1   \n",
       "26    BMDM_24hIL4.GSE159629.Mus_Musculus.H3K4me3.b2   \n",
       "27     BMDM_Basal.GSE159629.Mus_Musculus.H3K4me3.b1   \n",
       "28     BMDM_Basal.GSE159629.Mus_Musculus.H3K4me3.b2   \n",
       "29         3T3_Norm.GSE178344.Mus_Musculus.SMC1A.b1   \n",
       "30         3T3_TCF1.GSE178344.Mus_Musculus.SMC1A.b1   \n",
       "\n",
       "                               Control Used?  \n",
       "2   control present and used in peak calling  \n",
       "3   control present and used in peak calling  \n",
       "4                 no control for this sample  \n",
       "5                 no control for this sample  \n",
       "6                 no control for this sample  \n",
       "7                 no control for this sample  \n",
       "8   control present and used in peak calling  \n",
       "9   control present and used in peak calling  \n",
       "10  control present and used in peak calling  \n",
       "11  control present and used in peak calling  \n",
       "12  control present and used in peak calling  \n",
       "13  control present and used in peak calling  \n",
       "14                no control for this sample  \n",
       "15                no control for this sample  \n",
       "16  control present and used in peak calling  \n",
       "17  control present and used in peak calling  \n",
       "18  control present and used in peak calling  \n",
       "19  control present and used in peak calling  \n",
       "20                no control for this sample  \n",
       "21                no control for this sample  \n",
       "22                no control for this sample  \n",
       "23                no control for this sample  \n",
       "24  control present and used in peak calling  \n",
       "25  control present and used in peak calling  \n",
       "26  control present and used in peak calling  \n",
       "27  control present and used in peak calling  \n",
       "28  control present and used in peak calling  \n",
       "29                no control for this sample  \n",
       "30                no control for this sample  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "human_path = \"../../../results/samplesheets/fastq/chipseq_human_tracker.tsv\"\n",
    "mouse_path = \"../../../results/samplesheets/fastq/chipseq_mouse_tracker.tsv\"\n",
    "\n",
    "table = []\n",
    "df = generate_summary_df(mouse_path)\n",
    "#human_samples = get_sample_names(human_path)\n",
    "\n",
    "columns = [\"Sample Name\", \"Control Used?\"]\n",
    "df.columns = columns\n",
    "\n",
    "# reset the row count\n",
    "df = df.reset_index(drop=True)\n",
    "df.index += 2\n",
    "\n",
    "display(df)\n",
    "\n",
    "# samples = pd.concat([human_samples, mouse_samples])\n",
    "# samples = samples.dropna(how='all', subset=['Sample Name'])\n",
    "# samples = samples.reset_index(drop=True)\n",
    "\n",
    "# display(samples)\n",
    "\n",
    "# with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "#     display(samples)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.10",
   "language": "python",
   "name": "geo-parser"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "93ad387fa7c8f05260498cf670de53853982729883375a2905828e779a517e68"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
