{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import string\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many reads\n",
    "# whether there was a control or not and whether it is used in peak calling\n",
    "# how many peaks at what setting and threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_summary_df(path):\n",
    "    samples = []\n",
    "    with open(path,'r') as file:\n",
    "\n",
    "        #reader = csv.reader(file, delimiter='\\t')\n",
    "        #next(reader, None)\n",
    "        reader = csv.DictReader(file, delimiter='\\t')\n",
    "        for line in reader:\n",
    "            sample_info = []\n",
    "            \n",
    "            # if there is no sample name, skip this line\n",
    "            if not line['Sample Name']:\n",
    "                continue\n",
    "\n",
    "            ##### CONSTRUCT THE SAMPLE NAME ######\n",
    "\n",
    "            name = line['Sample Name']\n",
    "            gse = line['GSE ID of Corresponding HiChIP Data']\n",
    "\n",
    "            organism = line['Organism']\n",
    "            organism = string.capwords(organism).replace(\" \", \"_\")\n",
    "\n",
    "            target_of_antibody = line['ChIP-seq Pull Down']\n",
    "            if (target_of_antibody==\"N/A\"):\n",
    "                continue\n",
    "            \n",
    "            replicate = line['Replicate Serial No']\n",
    "\n",
    "            sample_name = name + \".\" + gse + \".\" + organism + \".\" + target_of_antibody + \".b\" + replicate\n",
    "\n",
    "            ###########\n",
    "\n",
    "            # determine whether a control is present\n",
    "            control_info = \"no control for this sample\"\n",
    "            if line['Control GSM(s)']!=\"N/A\":\n",
    "                control_info = \"control present and used in peak calling\"\n",
    "\n",
    "            \n",
    "            # determine peak sizes and settings\n",
    "            narrowpeak1 = \"peak calling not completed\"\n",
    "            narrowpeak2 = \"peak calling not completed\"\n",
    "\n",
    "            idrpeak1 = \"peak calling not completed\"\n",
    "            idrpeak2 = \"peak calling not completed\"\n",
    "\n",
    "            # check for merged replicates number of peaks\n",
    "            if line['Merged replicates (if there are replicates)']==\"1\":\n",
    "                merged_sample_name = sample_name[:-1] + \"1\"\n",
    "                idrpeakpath = \"../../../results/peaks/merged_chipline/\" + merged_sample_name\n",
    "                idrpeakpath1 = idrpeakpath + \"/FINAL_IDR_Peaks_FDR0.1.txt\"\n",
    "                idrpeakpath2 = idrpeakpath + \"/FINAL_IDR_Peaks_FDR0.05.txt\"\n",
    "\n",
    "                if os.path.exists(idrpeakpath):\n",
    "                    idrpeak1 = sum(1 for line in open(idrpeakpath1))\n",
    "                    idrpeak2 = sum(1 for line in open(idrpeakpath2))\n",
    "                else:\n",
    "                    idrpeak1 = \"N/A\"\n",
    "                    idrpeak2 = \"N/A\"\n",
    "            \n",
    "            # check for peaks per replicate\n",
    "            if line['Ran peak calling']==\"1\":\n",
    "                narrowpeakpath = \"../../../results/peaks/chipline/\" + sample_name + \"/MACS2_Ext*/\"\n",
    "                narrowpeakpath1 = narrowpeakpath + sample_name + \".macs2_peaks.narrowPeak_Q0.05filt\"\n",
    "                narrowpeakpath2 = narrowpeakpath + sample_name + \".macs2_peaks.narrowPeak_Q0.01filt\"\n",
    "\n",
    "                if len(glob.glob(narrowpeakpath1))==1:\n",
    "                    narrowpeak1 = sum(1 for line in open(glob.glob(narrowpeakpath1)[0]))\n",
    "                else:\n",
    "                    narrowpeak1 = \"error in peak calling\"\n",
    "\n",
    "                \n",
    "                if len(glob.glob(narrowpeakpath2))==1:\n",
    "                    narrowpeak2 = sum(1 for line in open(glob.glob(narrowpeakpath2)[0]))\n",
    "                else:\n",
    "                    narrowpeak2 = \"error in peak calling\"\n",
    "\n",
    "\n",
    "            sample_info = [sample_name, control_info, narrowpeak1, narrowpeak2, idrpeak1, idrpeak2]\n",
    "            samples.append(sample_info)\n",
    "    \n",
    "    df = pd.DataFrame(samples)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Control Used?</th>\n",
       "      <th>Narrow peaks with Q-value threshold of 0.01</th>\n",
       "      <th>Narrow peaks with Q-value threshold of 0.05</th>\n",
       "      <th>IDR peaks with respect to FDR 0.1</th>\n",
       "      <th>IDR peaks with respect to FDR 0.05</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EryP.GSE112717.Mus_Musculus.Gata1.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "      <td>152539</td>\n",
       "      <td>146931</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EryD.GSE112717.Mus_Musculus.Gata1.b1</td>\n",
       "      <td>control present and used in peak calling</td>\n",
       "      <td>50066</td>\n",
       "      <td>48138</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>180440</td>\n",
       "      <td>180440</td>\n",
       "      <td>23118</td>\n",
       "      <td>18701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>207431</td>\n",
       "      <td>203757</td>\n",
       "      <td>23118</td>\n",
       "      <td>18701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>241450</td>\n",
       "      <td>238302</td>\n",
       "      <td>24860</td>\n",
       "      <td>19835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>DCM_10.GSE165303.Homo_Sapiens.H3K27ac.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>124318</td>\n",
       "      <td>124318</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>hCM-Ctrl.GSE165303.Homo_Sapiens.H3K27ac.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>98693</td>\n",
       "      <td>98529</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>hCM-Ctrl.GSE165303.Homo_Sapiens.H3K27ac.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>92482</td>\n",
       "      <td>92332</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>hCM-HAND1OE.GSE165303.Homo_Sapiens.H3K27ac.b1</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>87092</td>\n",
       "      <td>87092</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>hCM-HAND1OE.GSE165303.Homo_Sapiens.H3K27ac.b2</td>\n",
       "      <td>no control for this sample</td>\n",
       "      <td>103869</td>\n",
       "      <td>103647</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Sample Name  \\\n",
       "2              EryP.GSE112717.Mus_Musculus.Gata1.b1   \n",
       "3              EryD.GSE112717.Mus_Musculus.Gata1.b1   \n",
       "4    AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b1   \n",
       "5    AML12_shCtrl.GSE141113.Mus_Musculus.H3K9me3.b2   \n",
       "6    AML12_shSafb.GSE141113.Mus_Musculus.H3K9me3.b1   \n",
       "..                                              ...   \n",
       "211        DCM_10.GSE165303.Homo_Sapiens.H3K27ac.b1   \n",
       "212      hCM-Ctrl.GSE165303.Homo_Sapiens.H3K27ac.b1   \n",
       "213      hCM-Ctrl.GSE165303.Homo_Sapiens.H3K27ac.b2   \n",
       "214   hCM-HAND1OE.GSE165303.Homo_Sapiens.H3K27ac.b1   \n",
       "215   hCM-HAND1OE.GSE165303.Homo_Sapiens.H3K27ac.b2   \n",
       "\n",
       "                                Control Used?  \\\n",
       "2    control present and used in peak calling   \n",
       "3    control present and used in peak calling   \n",
       "4                  no control for this sample   \n",
       "5                  no control for this sample   \n",
       "6                  no control for this sample   \n",
       "..                                        ...   \n",
       "211                no control for this sample   \n",
       "212                no control for this sample   \n",
       "213                no control for this sample   \n",
       "214                no control for this sample   \n",
       "215                no control for this sample   \n",
       "\n",
       "    Narrow peaks with Q-value threshold of 0.01  \\\n",
       "2                                        152539   \n",
       "3                                         50066   \n",
       "4                                        180440   \n",
       "5                                        207431   \n",
       "6                                        241450   \n",
       "..                                          ...   \n",
       "211                                      124318   \n",
       "212                                       98693   \n",
       "213                                       92482   \n",
       "214                                       87092   \n",
       "215                                      103869   \n",
       "\n",
       "    Narrow peaks with Q-value threshold of 0.05  \\\n",
       "2                                        146931   \n",
       "3                                         48138   \n",
       "4                                        180440   \n",
       "5                                        203757   \n",
       "6                                        238302   \n",
       "..                                          ...   \n",
       "211                                      124318   \n",
       "212                                       98529   \n",
       "213                                       92332   \n",
       "214                                       87092   \n",
       "215                                      103647   \n",
       "\n",
       "    IDR peaks with respect to FDR 0.1 IDR peaks with respect to FDR 0.05  \n",
       "2                                 N/A                                N/A  \n",
       "3                                 N/A                                N/A  \n",
       "4                               23118                              18701  \n",
       "5                               23118                              18701  \n",
       "6                               24860                              19835  \n",
       "..                                ...                                ...  \n",
       "211                               N/A                                N/A  \n",
       "212                               N/A                                N/A  \n",
       "213                               N/A                                N/A  \n",
       "214                               N/A                                N/A  \n",
       "215                               N/A                                N/A  \n",
       "\n",
       "[214 rows x 6 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "human_path = \"../../../results/samplesheets/fastq/chipseq_human_tracker.tsv\"\n",
    "mouse_path = \"../../../results/samplesheets/fastq/chipseq_mouse_tracker.tsv\"\n",
    "\n",
    "table = []\n",
    "mouse_df = generate_summary_df(mouse_path)\n",
    "human_df = generate_summary_df(human_path)\n",
    "\n",
    "df = pd.concat([mouse_df, human_df])\n",
    "\n",
    "columns = [\"Sample Name\", \"Control Used?\"]\n",
    "columns.extend([\"Narrow peaks with Q-value threshold of 0.01\", \"Narrow peaks with Q-value threshold of 0.05\"])\n",
    "columns.extend([\"IDR peaks with respect to FDR 0.1\", \"IDR peaks with respect to FDR 0.05\"])\n",
    "df.columns = columns\n",
    "\n",
    "# reset the row count\n",
    "df = df.reset_index(drop=True)\n",
    "df.index += 2\n",
    "\n",
    "display(df)\n",
    "\n",
    "# samples = pd.concat([human_samples, mouse_samples])\n",
    "# samples = samples.dropna(how='all', subset=['Sample Name'])\n",
    "# samples = samples.reset_index(drop=True)\n",
    "\n",
    "# display(samples)\n",
    "\n",
    "# with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "#     display(samples)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.10",
   "language": "python",
   "name": "geo-parser"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "93ad387fa7c8f05260498cf670de53853982729883375a2905828e779a517e68"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
